File: pct-id-benchmark/scripts/boio/fasta.py
from dataclasses import dataclass, field
from typing import Dict


@dataclass
class FastaRecord:
    name: str
    extra: str
    sequence: str

    def __str__(self) -> str:
        lines = [f">{self.name} {self.extra}"]
        lines += [self.sequence[i:i + 80] for i in range(0, len(self.sequence), 80)]
        lines += [""]
        return "\n".join(lines)


@dataclass
class Fasta:
    records: Dict[str, FastaRecord] = field(default_factory=dict)

    @staticmethod
    def from_path(path: str) -> "Fasta":
        fasta = Fasta()
        with open(path, "r") as f:
            header = None
            seq_chunks = []
            for line in f:
                line = line.strip()
                if not line:
                    continue
                if line.startswith(">"):
                    if header:
                        parts = header.split(maxsplit=1)
                        name = parts[0]
                        extra = parts[1] if len(parts) > 1 else ""
                        fasta.records[name] = FastaRecord(name, extra, ''.join(seq_chunks))
                    header = line[1:]
                    seq_chunks = []
                else:
                    seq_chunks.append(line)

            if header:
                parts = header.split(maxsplit=1)
                name = parts[0]
                extra = parts[1] if len(parts) > 1 else ""
                fasta.records[name] = FastaRecord(name, extra, ''.join(seq_chunks))

        return fasta

    def write(self, file):
        for rec in self.records.values():
            file.write(str(rec))

    def __iter__(self):
        return iter(self.records.values())

File: rust/bioio/src/fasta.rs
use std::{
    fmt::Display,
    io::{BufRead, BufReader, Read},
    path::Path,
};

use anyhow::Context;
use indexmap::IndexMap;

#[derive(Default, Clone, PartialEq)]
pub struct FastaRecord {
    pub name: String,
    pub extra: String,
    pub sequence: String,
}

impl Display for FastaRecord {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        writeln!(f, ">{} {}", self.name, self.extra)?;
        let mut chunks = self
            .sequence
            .as_bytes()
            .chunks(60)
            .map(|c| std::str::from_utf8(c).unwrap());

        if let Some(last) = chunks.next_back() {
            chunks.try_for_each(|c| writeln!(f, "{c}"))?;
            write!(f, "{last}")
        } else {
            Ok(())
        }
    }
}

#[derive(Default)]
pub struct Fasta {
    pub records: IndexMap<String, FastaRecord>,
}

impl Fasta {
    pub fn parse<R: Read>(buf: R) -> anyhow::Result<Self> {
        let reader = BufReader::new(buf);

        let mut fasta = Self::default();
        let mut rec = FastaRecord::default();
        for line in reader.lines() {
            let line = line?;
            if let Some(line) = line.strip_prefix(">") {
                if !rec.name.is_empty() {
                    fasta.records.insert(rec.name.to_string(), rec);
                }
                rec = FastaRecord::default();

                let tokens = line.splitn(2, char::is_whitespace).collect::<Vec<_>>();
                rec.name = tokens[0].to_string();
                rec.extra = tokens[1].to_string();
            } else {
                rec.sequence.push_str(&line)
            }
        }
        fasta.records.insert(rec.name.to_string(), rec);

        Ok(fasta)
    }

    pub fn from_path<P: AsRef<Path>>(path: P) -> anyhow::Result<Self> {
        let f = std::fs::File::open(path.as_ref())?;
        Self::parse(f).with_context(|| format!("failed to parse Fasta from: {:?}", path.as_ref()))
    }
}
